### DISCUSSION TITLE
column.fill or nested fill issue

### DISCUSSION THREAD

**Discussion by oOBlinkOo** [exceljs/exceljs#1718] [Q&A] [answered]
For my columns (headers) im trying to set a specific color to each column.
```js
    let columns=[
        {header:"ID",key:"ID"},
        {header:"BEST_COLUMN",key:"BEST_COLUMN",width:30 },
        {header:"TEST_COLUMN",key:"TEST_COLUMN",width:30},

    ];
```
this is what I have for the creation of the headers but the issue is when I want to set the color:
color.

```js
  addColors(columnName,worksheet){
    let column:Column=worksheet.getColumn(columnName);
            if(column!=null){         
              let fill:Fill= {
                          type: 'pattern',
                          pattern: 'solid',
                          fgColor: { argb:  COLOR.GRAY}
                      };
              column.style = JSON.parse(JSON.stringify(column.style)); //very bad way to clone anythink, but works.
              column.fill = fill;
            }
  }
```
if the columns share the same **width** and it is **next to** the column that should change his color, that column will change the background of the column  to gray too (TEST_COLUMN).
In the next img only the columns with ID and BEST_COLUMN should be change to gray.
![image](https://user-images.githubusercontent.com/5222538/119188739-b4dd3000-ba40-11eb-8020-a58e745db792.png)

I was thinking that it was something related with:
https://github.com/exceljs/exceljs/issues/791
https://github.com/exceljs/exceljs/pull/1573
but I already try JSON.parse and it is still breaking my file.

If i dont set any width, my functions works perfectly

**Comment by Siemienik** [comment, marked as answer]
Hi, this a bug.

Following the source code:
https://github.com/exceljs/exceljs/blob/3b261536dc81f0688480b04df920055ab4e2f488/lib/doc/column.js#L206-L212

Even if you make a deep copy, that function applies style without deep copying. So, the behavior that you described is a bug and needs to be fixed. 

I accept orders for fixing and developing ExcelJS, to make an order please contact consulting@siemienik.pl

## Workaround

it might be easily workarounded by this code:

```ts
addColors(columnName,worksheet){
    let column:Column=worksheet.getColumn(columnName);
    if(column!=null){
        let fill:Fill= {
            type: 'pattern',
            pattern: 'solid',
            fgColor: { argb:  COLOR.GRAY}
        };
        const style = JSON.parse(JSON.stringify(column.style)); //very bad way to clone anythink, but works.

        style.fill = fill; 
        column.style = style
        column.eachCell(cell => {      cell.style = style;    });

    }
}
```

**Alternativelly**
btw. If you generate files with lots of styles I think you should check `xlsx-renderer` which based on ExcelJS: https://github.com/Siemienik/XToolset/tree/master/packages/xlsx-renderer#how-it-works
